controllers/helmrelease_controller_chart.go

```diff
@@ -126,13 +129,43 @@ func (r *HelmReleaseReconciler) loadHelmChart(source *sourcev1.HelmChart) (*char
 		return nil, fmt.Errorf("artifact '%s' download failed (status code: %s)", source.GetArtifact().URL, resp.Status)
 	}
 
-	if _, err = io.Copy(f, resp.Body); err != nil {
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download artifact, error: %w", err)
+	}
+
+	// verify checksum matches origin
+	if err := r.verifyArtifact(source.GetArtifact(), bytes.NewBuffer(bodyBytes)); err != nil {
+		return nil, err
+	}
+
+	if _, err = io.Copy(f, bytes.NewBuffer(bodyBytes)); err != nil {
```

Comment:
If you would use a `MultiWriter` in the `io.Copy` with a hasher, you wouldn't have to read all bytes, and wouldn't have to copy twice.